crates/libs/windows/src/core/borrowed.rs

```diff
@@ -1,50 +1,77 @@
+/// the following error, it's likely because you are trying to pass an owned value instead of a borrowed value:
 ///
-/// Many APIs take a params of the form `Into<Borrowed<'a, T>>`. This means you can pass anything that has an `Into` implementation 
-/// for `Borrowed<'a, T>`. Generally, if this is safe to do, the `windows` crate provides an implementation.
+/// ```
+///  the trait bound `SomeType: Into<Borrowed<'_, SomeType>>` is not satisfied
+/// ```
 #[repr(transparent)]
 pub struct Borrowed<'a, T> {
-	item: core::mem::ManuallyDrop<T>,
-    lifetime: core::marker::PhantomData<&'a ()>
+    item: Option<core::mem::ManuallyDrop<T>>,
+    lifetime: core::marker::PhantomData<&'a ()>,
 }
 
-impl <'a, T> Borrowed<'a, T> {
+impl<'a, T: super::Abi> Borrowed<'a, T> {
     /// Create a new `Borrowed` value.
     ///
     /// Normally, it is not necessary to use this function. Generally, there is a `From` implementation
     /// that allows you to call `.into` to safely create a `Borrowed` value.
-    ///
-	/// # Safety
-    ///
-    /// It must be safe to alias `T` as long as `T`'s destructor is not run. 
-    ///
-    /// For example, `IUnknown` is fine to alias as long as the reference count is guaranteed to stay above 0 
-    /// until the original `item` is dropped. On the other hand `Box<T>` is not fine to alias and thus passing a `Box` 
-    /// to this function would be UB.
-    unsafe fn new(item: &'a T) -> Self {
-        let item = unsafe { core::mem::transmute_copy(item) };
-        Self {
-            item,
-            lifetime: core::marker::PhantomData
+    pub fn new(item: Option<&'a T>) -> Self {
+        // SAFETY: `T` is safe to alias since that is a condition of implmenting `Abi`
+        let item = item.map(|i| core::mem::ManuallyDrop::new(unsafe { core::mem::transmute_copy(i) }));
+        Self { item, lifetime: core::marker::PhantomData }
+    }
+
+    /// Get the abi representation for this param
+    pub fn abi(&self) -> T::Abi {
+        match &self.item {
+            Some(inner) => inner.abi(),
+            // SAFETY: the `Abi` trait guarantees that all zeros is a valid representation
+            // of its associated `Abi` type.
+            None => unsafe { std::mem::MaybeUninit::zeroed().assume_init() },
         }
     }
 }
 
-impl<'a, T> std::ops::Deref for Borrowed<'a, T> {
-    type Target = T;
-    fn deref(&self) -> &Self::Target {
-        &*self.item
+impl<'a, T> Borrowed<'a, T> {
+    /// Get an optional reference to the underlying value
+    pub fn as_ref(&self) -> Option<&T> {
+        self.item.as_ref().map(|s| &**s)
+    }
+}
+
+impl <'a, T: 'a , U> From<&'a U> for Borrowed<'a, T> where T: super::Abi, &'a U: Into<&'a T> {
+    fn from(item: &'a U) -> Self {
+        Borrowed::new(Some(item.into()))
     }
-}
\ No newline at end of file
+}
+
+impl <'a, T: super::Abi> From<Option<&'a T>> for Borrowed<'a, T> where  &'a T: Into<Borrowed<'a, T>> {
```

Comment:
Typically, this is going to be `&Option<T>` rather than `Option<&T>`. Not sure if we need both but the former is definitely more useful in practice. e.g. 

https://github.com/microsoft/windows-rs/blob/11a2db2a161c5bef674573ac4e2121e8c045c708/crates/samples/direct2d/src/main.rs#L154-L159

Reply:
Getting the Direct2D sample to compile would be a good smoke test.

Reply:
Here are the changes to get d2d working: https://github.com/microsoft/windows-rs/pull/1811/commits/792a1e975a77f0261fc5bc457a58e7fe653cd3c8